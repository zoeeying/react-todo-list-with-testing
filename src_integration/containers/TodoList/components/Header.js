import React, { Component } from 'react'
import { connect } from 'react-redux'
import { actions } from '../store'
import '../index.css'

class Header extends Component {
  onInputKeyUp = e => {
    const { inputValue, addUndoItem, onInputChange } = this.props
    if (e.keyCode === 13 && !!inputValue) {
      addUndoItem(inputValue)
      onInputChange('')
    }
  }

  render() {
    const { inputValue, onInputChange } = this.props
    return (
      <div className="header">
        <div className="header_content">
          TodoList
          <input
            placeholder="Add Todo"
            className="header_input"
            data-test="header-input"
            value={inputValue}
            onChange={e => {
              onInputChange(e.target.value)
            }}
            onKeyUp={this.onInputKeyUp}
          ></input>
        </div>
      </div>
    )
  }
}

const mapState = state => ({
  inputValue: state.todo.inputValue,
})
const mapDispatch = dispatch => ({
  onInputChange(value) {
    dispatch(actions.changeInputValue(value))
  },
})

export default connect(mapState, mapDispatch)(Header)
