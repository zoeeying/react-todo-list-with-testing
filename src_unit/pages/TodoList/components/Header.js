import React, { Component } from 'react'
import '../index.css'

class Header extends Component {
  constructor(props) {
    super(props)
    this.state = {
      inputValue: '',
    }
  }

  onInputChange = e => {
    this.setState({
      inputValue: e.target.value,
    })
  }

  onInputKeyUp = e => {
    const { inputValue } = this.state
    const { addUndoItem } = this.props
    if (e.keyCode === 13 && !!inputValue) {
      addUndoItem(inputValue)
      this.setState({
        inputValue: '',
      })
    }
  }

  render() {
    const { inputValue } = this.state
    return (
      <div className="header">
        <div className="header_content">
          TodoList
          <input
            placeholder="Add Todo"
            className="header_input"
            data-test="header-input"
            value={inputValue}
            onChange={this.onInputChange}
            onKeyUp={this.onInputKeyUp}
          ></input>
        </div>
      </div>
    )
  }
}

export default Header
