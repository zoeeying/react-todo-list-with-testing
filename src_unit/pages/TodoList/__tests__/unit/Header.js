import React from 'react'
import { shallow } from 'enzyme'
import Header from '../../components/Header'
import { findTestWrapper } from '../../../../utils/testUtils'

describe('测试Header组件', () => {
  it('快照测试，保证样式渲染正常', () => {
    const wrapper = shallow(<Header />)
    expect(wrapper).toMatchSnapshot()
  })

  it('包含input', () => {
    const wrapper = shallow(<Header />)
    const inputElement = findTestWrapper(wrapper, 'header-input')
    expect(inputElement.length).toBe(1)
  })

  it('input初始值是空串', () => {
    const wrapper = shallow(<Header />)
    const inputElement = findTestWrapper(wrapper, 'header-input')
    expect(inputElement.prop('value')).toEqual('')
  })

  it('input的值会跟随用户输入变化', () => {
    const wrapper = shallow(<Header />)
    const inputElement = findTestWrapper(wrapper, 'header-input')
    const inputValue = '今天是好好学习的叙宝'
    // 写法1：对组件中的state数据进行测试，适用于单元测试
    inputElement.simulate('change', {
      // 模拟事件
      target: { value: inputValue },
    })
    expect(wrapper.state('inputValue')).toEqual(inputValue)
    // 写法2：对DOM的属性进行测试，适用于集成测试
    // 这里需要重新获取一下DOM元素，重新渲染后的DOM元素上的属性值才是更新后的值
    // const inputElement2 = wrapper.find("input[data-test='header-input']")
    // expect(inputElement2.prop('value')).toBe(inputValue)
  })

  it('input框回车，如果值为空，无操作', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Header addUndoItem={fn} />)
    const inputElement = findTestWrapper(wrapper, 'header-input')
    wrapper.setState({ inputValue: '' }) // input框没有值
    inputElement.simulate('keyUp', {
      keyCode: 13,
    })
    expect(fn).not.toHaveBeenCalled()
  })

  it('input框回车，如果值不为空，函数被调用，最后清除掉input框中的内容', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Header addUndoItem={fn} />)
    const inputElement = findTestWrapper(wrapper, 'header-input')
    wrapper.setState({ inputValue: '今天是好好学习的叙宝' }) // input框输入了值
    inputElement.simulate('keyUp', {
      keyCode: 13,
    })
    expect(fn).toHaveBeenCalled()
    expect(fn).toHaveBeenLastCalledWith('今天是好好学习的叙宝')
    const newInputElement = findTestWrapper(wrapper, 'header-input')
    expect(newInputElement.prop('value')).toBe('')
  })
})
