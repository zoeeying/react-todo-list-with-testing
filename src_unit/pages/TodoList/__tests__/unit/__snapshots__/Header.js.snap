// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试Header组件 快照测试，保证样式渲染正常 1`] = `
<div
  className="header"
>
  <div
    className="header_content"
  >
    TodoList
    <input
      className="header_input"
      data-test="header-input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Add Todo"
      value=""
    />
  </div>
</div>
`;
